File: src/derivatives-trading-portfolio-margin-pro-ws-streams.ts
import {
    DerivativesTradingPortfolioMarginPro,
    DERIVATIVES_TRADING_PORTFOLIO_MARGIN_PRO_WS_STREAMS_PROD_URL,
} from '@binance/derivatives-trading-portfolio-margin-pro';
import inquirer from 'inquirer';
import {
    decodeSelectedEntities,
    getConfigurationRestAPI,
    getUserAgent,
    isEmpty,
    readStdinObj,
    getParsedArgs,
} from './utils';

const parsedArgs = getParsedArgs();
const isFullDescription = parsedArgs?.['full-description'] ?? false;

const getClient = () => {
    if (!process.env?.LOG_LEVEL) {
        process.env.LOG_LEVEL = 'ERROR';
    }
    process.env.BINANCE_CONNECTOR_JS_USER_AGENT = getUserAgent(
        'derivatives-trading-portfolio-margin-pro'
    );

    let basePath = DERIVATIVES_TRADING_PORTFOLIO_MARGIN_PRO_WS_STREAMS_PROD_URL;

    const configurationRestAPI = getConfigurationRestAPI(
        parsedArgs?.profile,
        'derivatives-portfolio-margin-pro'
    );

    if (process.env.BINANCE_DERIVATIVES_PORTFOLIO_MARGIN_PRO_WS_STREAMS_BASE_PATH) {
        basePath = process.env.BINANCE_DERIVATIVES_PORTFOLIO_MARGIN_PRO_WS_STREAMS_BASE_PATH;
    } else if (configurationRestAPI && configurationRestAPI['wsStreamsBasePath']) {
        basePath = configurationRestAPI['wsStreamsBasePath'];
    }

    const configurationWebsocketStreams = {
        wsURL: basePath,
    };

    return {
        client: new DerivativesTradingPortfolioMarginPro({ configurationWebsocketStreams }),
    };
};
const derivativesTradingPortfolioMarginProCommands: any[] = [];

derivativesTradingPortfolioMarginProCommands.push({
    command: 'user-data',
    describe: decodeSelectedEntities(
        'Subscribes to the user data WebSocket stream using the provided listen key.',
        isFullDescription
    ),
    builder: (yargsCmd: any) => {
        return yargsCmd
            .options({
                'listen-key': {
                    type: 'string',
                    group: 'Command Options:',
                },
                id: {
                    type: 'string',
                    group: 'Command Options:',
                },
                json: {
                    describe: 'Send all fields as JSON',
                    type: 'string',
                    group: 'JSON Options:',
                },
            })
            .check((options: any) => {
                const requiredParams: any = [];
                const stdinObj: any = readStdinObj();

                if (!isEmpty(stdinObj)) {
                    options = { ...options, ...stdinObj };
                }

                if (options.json) {
                    options = { ...options, ...JSON.parse(options.json) };
                }

                if (!options?.['listen-key'] && !options?.interactive) {
                    requiredParams.push('listen-key');
                }

                if (requiredParams.length > 0) {
                    return `Following arguments are required: ${requiredParams.join(', ')}`;
                }

                return true;
            });
    },
    handler: async (options: any) => {
        const questions: any = [];
        const stdinObj: any = readStdinObj();

        if (!isEmpty(stdinObj)) {
            options = { ...options, ...stdinObj };
        }

        if (options.json) {
            options = { ...options, ...JSON.parse(options.json) };
            delete options.json;
        }

        if (options.interactive && !options?.['listen-key']) {
            questions.push({
                type: 'input',
                name: 'listen-key',
                message: 'Input listenKey:',
                validate: (input: string) => (input ? true : 'listen-key cannot be empty'),
            });
        }

        if (questions.length > 0) {
            const answers = await inquirer.prompt(questions);
            options = { ...options, ...answers };
        }

        try {
            const { client } = getClient();
            const connection = await client.websocketStreams.connect();
            const stream = connection.userData(options?.['listen-key'], options?.['id']);
            const limit = options?.['stream-limit'];
            const startTime = Date.now();
            const durationMs = options?.['stream-duration'];
            let eventCount = 0;
            stream.on('message', (data: any) => {
                if (options?.['pretty']) {
                    console.info(data);
                } else {
                    console.info(JSON.stringify(data));
                }
                if (eventCount > limit || Date.now() - startTime > durationMs) {
                    connection.disconnect();
                    return;
                }
                eventCount = eventCount + 1;
            });
        } catch (e: any) {
            console.log(e.message);
            return;
        }
    },
});

export default {
    command: 'derivatives-portfolio-margin-pro-streams',
    description: 'Binance Derivatives Trading Portfolio Margin Pro WebSocket Market Streams',
    builder: (yargs: any) => {
        yargs
            .usage(
                'Usage: binance-cli derivatives-portfolio-margin-pro-streams <command> [options]'
            )
            .options({
                pretty: {
                    type: 'boolean',
                    group: 'Stream Options:',
                    describe: 'Pretty-print the JSON output',
                },
                'stream-limit': {
                    type: 'number',
                    group: 'Stream Options:',
                    describe: 'Stop listening after the given number of messages',
                },
                'stream-duration': {
                    type: 'number',
                    group: 'Stream Options:',
                    describe: 'Stop listening after the given duration (in ms)',
                },
            });
        derivativesTradingPortfolioMarginProCommands.forEach((command: any) => {
            yargs.command(command);
        });
    },
};

File: src/derivatives_trading_portfolio_margin_pro_ws_streams.rs
use crate::utils::{
    decode_selected_entities, get_client_configuration, init_user_agent, wait_for_shutdown,
};
use binance_sdk::config::ConfigurationWebsocketStreams;
use binance_sdk::constants::DERIVATIVES_TRADING_PORTFOLIO_MARGIN_PRO_WS_STREAMS_PROD_URL;
use binance_sdk::derivatives_trading_portfolio_margin_pro::DerivativesTradingPortfolioMarginProWsStreams;
use binance_sdk::derivatives_trading_portfolio_margin_pro::websocket_streams::WebsocketStreamsHandle;
use clap::Subcommand;
use dialoguer::Input;
use std::env;
use std::io::{Error, ErrorKind};

fn get_client(profile: Option<&str>) -> Result<WebsocketStreamsHandle, Error> {
    init_user_agent("derivatives-trading-portfolio-margin-pro");

    let client_config =
        get_client_configuration(profile, "derivatives-trading-portfolio-margin-pro").unwrap();
    let api_env = env::var("BINANCE_API_ENV")
        .ok()
        .or(client_config.env)
        .unwrap_or_else(|| "prod".to_string());

    let base_path = client_config
        .base_path_ws_streams
        .unwrap_or(match api_env.as_str() {
            "prod" => DERIVATIVES_TRADING_PORTFOLIO_MARGIN_PRO_WS_STREAMS_PROD_URL.to_string(),
            _ => {
                return Err(Error::new(
                    ErrorKind::InvalidInput,
                    "Invalid api env, valid values: prod",
                ));
            }
        });

    let builder = ConfigurationWebsocketStreams::builder().ws_url(base_path);

    let ws_config = builder
        .build()
        .map_err(|err| Error::new(ErrorKind::Other, err))?;

    Ok(DerivativesTradingPortfolioMarginProWsStreams::from_config(
        ws_config,
    ))
}

#[derive(Subcommand)]
pub enum DerivativesTradingPortfolioMarginProWebsocketStreamsCommands {
    #[command(
        about = decode_selected_entities(r#"Subscribes to the user data WebSocket stream using the provided listen key."#, false),
    )]
    UserData(UserDataArgs),
}

pub async fn handle_derivatives_trading_portfolio_margin_pro_ws_streams_command(
    command: DerivativesTradingPortfolioMarginProWebsocketStreamsCommands,
) -> anyhow::Result<()> {
    match command {
        DerivativesTradingPortfolioMarginProWebsocketStreamsCommands::UserData(args) => {
            user_data(args).await
        }
    }
}

#[derive(clap::Parser, Debug)]
pub struct UserDataArgs {
    #[arg(long)]
    pub profile: Option<String>,

    #[arg(long = "listen-key")]
    pub listen_key: Option<String>,

    #[arg(long)]
    pub id: Option<String>,

    /// Send all fields as JSON
    #[arg(long)]
    pub json: Option<String>,

    #[arg(long)]
    pub interactive: bool,

    /// Pretty-print the JSON output
    #[arg(long)]
    pub pretty: bool,

    /// Stop listening after the given number of messages
    #[arg(long = "stream-limit")]
    pub stream_limit: Option<u64>,

    /// Stop listening after the given duration in ms
    #[arg(long = "stream-duration")]
    pub stream_duration: Option<u64>,
}

async fn user_data(mut args: UserDataArgs) -> anyhow::Result<()> {
    let rest_client = get_client(args.profile.as_deref())?;

    if args.interactive {
        if args.listen_key.is_none() {
            let listen_key: String = Input::new()
                .with_prompt("Input listenKey")
                .interact_text()?;

            args.listen_key = Some(listen_key);
        }
    }

    let connection = rest_client.connect().await?;

    // Make the API call
    let stream = connection
        .user_data(
            args.listen_key
                .ok_or_else(|| anyhow::anyhow!("listen_key is required"))?,
            args.id,
        )
        .await?;

    let message_count = std::sync::Arc::new(std::sync::atomic::AtomicU64::new(0));
    let stop = std::sync::Arc::new(tokio::sync::Notify::new());

    let pretty = args.pretty;
    let stream_limit = args.stream_limit;
    let stream_duration = args.stream_duration;

    let message_count_for_handler = std::sync::Arc::clone(&message_count);
    let stop_for_handler = std::sync::Arc::clone(&stop);

    stream.on_message(move |data| {
        let output = if pretty {
            serde_json::to_string_pretty(&data)
        } else {
            serde_json::to_string(&data)
        };

        println!("{}", output.unwrap());

        if let Some(limit) = stream_limit {
            let count =
                message_count_for_handler.fetch_add(1, std::sync::atomic::Ordering::Relaxed) + 1;

            if count >= limit {
                stop_for_handler.notify_one();
            }
        }
    });

    if stream_limit == Some(0) {
        stop.notify_one();
    }

    if let Some(ms) = stream_duration {
        tokio::select! {
            _ = stop.notified() => {
                eprintln!("stream limit reached");
            }
            _ = tokio::time::sleep(std::time::Duration::from_millis(ms)) => {
                eprintln!("stream duration reached");
            }
            _ = wait_for_shutdown() => {
                eprintln!("shutdown requested");
            }
        }
    } else {
        tokio::select! {
            _ = stop.notified() => {
                eprintln!("stream limit reached");
            }
            _ = wait_for_shutdown() => {
                eprintln!("shutdown requested");
            }
        }
    }

    Ok(())
}
